common/promise_cache/src/lib.rs

```diff
@@ -1,227 +1,120 @@
-        let target_cache_size = self.capacity.saturating_sub(1);
-        if let Some(prune_count) = self.cache.len().checked_sub(target_cache_size) {
-            let keys_to_prune = self
-                .cache
-                .keys()
-                .filter(|k| !self.protector.protect_from_eviction(*k))
-                .sorted_by_key(|k| self.protector.sort_key(k))
-                .take(prune_count)
-                .cloned()
-                .collect::<Vec<_>>();
-
-            for key in &keys_to_prune {
-                self.protector.notify_eviction(key, &self.logger);
-                self.cache.remove(key);
+    /// If the closure is successful, the computed value is returned. Otherwise, a
+    /// [`PromiseCacheError`] is returned.
+    ///
+    /// The result values are not retained: as soon as the first thread has returned, new threads
+    /// will recompute the value again. Therefore, you should store the resulting value in another
+    /// cache, so that threads that are just a bit too late can still use the value computed herein.
+    ///
+    /// It is possible (and in some cases, advisable) to provide different closures at different
+    /// code locations for the same `PromiseCache`: If computation is easier in some contexts,
+    /// other threads also may also benefit from that. However, if a thread calls `get_or_compute`
+    /// with a "fast" closure while computation is already in progress with a "slow" closure, that
+    /// thread may wait longer than it would have by simply using its "fast" closure. This is
+    /// unavoidable as we can not compute the complexity of closures.
+    ///
+    /// NOTE: do not hold any locks while calling this function! Lock necessary locks within the
+    /// passed closure instead.
+    pub fn get_or_compute<F, E>(&self, key: &K, computation: F) -> Result<V, PromiseCacheError<E>>
+    where
+        F: FnOnce() -> Result<V, E>,
+    {
+        let mut cache = self.cache.lock();
+        match cache.get(key) {
+            Some(item) => {
+                let item = item.clone();
+                drop(cache);
+                item.recv()
+                    .map_err(|_| PromiseCacheError::Panic)
+                    .and_then(|res| res.map_err(|_| PromiseCacheError::Error(None)))
+            }
+            None => {
+                let (sender, receiver) = oneshot();
+                cache.insert(key.clone(), receiver);
+                drop(cache);
+                match computation() {
+                    Ok(value) => {
+                        sender.send(Ok(value));
+                        Ok(self
+                            .cache
+                            .lock()
+                            .remove(key)
+                            // PANIC: should not happen, as the insert and remove is guarded so that
+                            // no two threads will end up removing the same key.
+                            .expect("value has vanished")
```

Comment:
Rather than `expect`s I think it would be better to have a `PromiseCacheError::LogicError` variant. Seeing as we're already returning a result the only concern is the perf penalty which I think will be minimal.